sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowGroupLimitExec.scala

```diff
@@ -0,0 +1,251 @@
+  def increaseRank(): Unit
+
+  // Clear the rank value.
+  def clearRank(): Unit
+
+  var bufferIterator: Iterator[InternalRow] = _
+
+  private[this] def fetchNextPartition(): Unit = {
+    clearRank()
+    bufferIterator = createGroupIterator()
+  }
+
+  override final def hasNext: Boolean =
+    (bufferIterator != null && bufferIterator.hasNext) || nextRowAvailable
+
+  override final def next(): InternalRow = {
+    // Load the next partition if we need to.
+    if ((bufferIterator == null || !bufferIterator.hasNext) && nextRowAvailable) {
+      fetchNextPartition()
+    }
+
+    if (bufferIterator.hasNext) {
+      bufferIterator.next()
+    } else {
+      throw new NoSuchElementException
+    }
+  }
+
+  private def createGroupIterator(): Iterator[InternalRow] = {
+    new Iterator[InternalRow] {
+      // Before we start to fetch new input rows, make a copy of nextGroup.
+      val currentGroup = nextGroup.copy()
+
+      def hasNext: Boolean = {
+        if (nextRowAvailable) {
+          if (exceedingLimit() && nextGroup == currentGroup) {
+            do {
+              fetchNextRow()
+            } while (nextRowAvailable && nextGroup == currentGroup)
+          }
+          nextRowAvailable && nextGroup == currentGroup
+        } else {
+          nextRowAvailable
+        }
+      }
+
+      def next(): InternalRow = {
+        val currentRow = nextRow.copy()
+        increaseRank()
+        fetchNextRow()
+        currentRow
+      }
+    }
+  }
+}
+
+case class SimpleGroupLimitIterator(
+    partitionSpec: Seq[Expression],
+    output: Seq[Attribute],
+    stream: Iterator[InternalRow],
+    limit: Int) extends WindowIterator {
+  var count = 0
+
+  override def exceedingLimit(): Boolean = {
+    count >= limit
+  }
+
+  override def increaseRank(): Unit = {
+    count += 1
+  }
+
+  override def clearRank(): Unit = {
+    count = 0
+  }
+}
+
+case class RankGroupLimitIterator(
+    partitionSpec: Seq[Expression],
+    output: Seq[Attribute],
+    stream: Iterator[InternalRow],
+    orderSpec: Seq[SortOrder],
+    limit: Int) extends WindowIterator {
+  val ordering = GenerateOrdering.generate(orderSpec, output)
+  var count = 0
+  var rank = 0
+  var currentRank: UnsafeRow = null
+
+  override def exceedingLimit(): Boolean = {
+    rank >= limit
+  }
+
+  override def increaseRank(): Unit = {
+    if (count == 0) {
+      currentRank = nextRow.copy()
+    } else {
+      if (ordering.compare(currentRank, nextRow) != 0) {
+        rank = count
+        currentRank = nextRow.copy()
+      }
+    }
+    count += 1
+  }
+
+  override def clearRank(): Unit = {
+    count = 0
+    rank = 0
+    currentRank = null
+  }
+}
+
+case class DenseRankGroupLimitIterator(
+    partitionSpec: Seq[Expression],
+    output: Seq[Attribute],
+    stream: Iterator[InternalRow],
+    orderSpec: Seq[SortOrder],
+    limit: Int) extends WindowIterator {
+  val ordering = GenerateOrdering.generate(orderSpec, output)
+  var rank = 0
```

Comment:
We can move `var rand = 0` to the parent class, then we don't need to add `def exceedingLimit`

Reply:
For `SimpleGroupLimitIterator`, rank is simply the count.